File: src/lib/repositories/base-repository.ts
import { Pool } from 'pg';
import pool from '../db'
import { QueryBuilder } from '../db/query-builder';
import { BaseEntity, QueryParams, PaginatedResult } from '../models/types';

/**
 * Base repository class that implements common CRUD operations
 * @template T The entity type
 */
export abstract class BaseRepository<T extends BaseEntity> {
  protected pool: Pool;
  protected tableName: string;

  /**
   * Create a new repository instance
   * @param tableName The database table name
   * @param pool Optional database pool (uses default if not provided)
   */
  constructor(tableName: string, other_pool?: Pool) {
    this.pool = pool || other_pool;
    this.tableName = tableName;
  }

  /**
   * Create a new query builder instance
   */
  protected createQueryBuilder(): QueryBuilder {
    return new QueryBuilder(this.tableName, this.pool);
  }

  /**
   * Find an entity by ID
   * @param id Entity ID
   */
  async findById(id: number): Promise<T | null> {
    return this.createQueryBuilder()
      .where('id = $1', id)
      .executeSingle<T>();
  }

  /**
   * Find all entities
   * @param params Optional query parameters
   */
  async findAll(params?: QueryParams): Promise<T[]> {
    const builder = this.createQueryBuilder();
    
    if (params) {
      builder.applyQueryParams(params);
    }
    
    return builder.execute<T>();
  }

  /**
   * Find entities with pagination
   * @param params Query parameters
   */
  async findPaginated(params: QueryParams): Promise<PaginatedResult<T>> {
    const page = params.page || 1;
    const limit = params.limit || 20;
    
    const builder = this.createQueryBuilder();
    builder.applyQueryParams(params);
    
    return builder.executePaginated<T>(page, limit);
  }

  /**
   * Create a new entity
   * @param data Entity data
   */
  async create(data: Partial<T>): Promise<T> {
    return this.createQueryBuilder().insert<T>(data);
  }

  /**
   * Update an existing entity
   * @param id Entity ID
   * @param data Updated entity data
   */
  async update(id: number, data: Partial<T>): Promise<T | null> {
    return this.createQueryBuilder().update<T>(id, data);
  }

  /**
   * Delete an entity
   * @param id Entity ID
   */
  async delete(id: number): Promise<boolean> {
    return this.createQueryBuilder().delete(id);
  }

  /**
   * Count entities matching the given criteria
   * @param params Query parameters
   */
  async count(params?: QueryParams): Promise<number> {
    const builder = this.createQueryBuilder().select(['COUNT(*) as count']);
    
    if (params) {
      builder.applyQueryParams(params);
    }
    
    const result = await builder.executeSingle<{ count: string }>();
    return result ? parseInt(result.count, 10) : 0;
  }
}

File: venue-explorer-api/app/repositories/base_repository.py
"""
Base repository class with common database operations
"""

from sqlalchemy.orm import Session
from typing import TypeVar, Generic, Optional, List, Type, Any
from app.models.orm import Base

T = TypeVar('T', bound=Base)


class BaseRepository(Generic[T]):
    """Base repository with common CRUD operations"""
    
    def __init__(self, db: Session, model: Type[T]):
        self.db = db
        self.model = model
    
    async def find_by_id(self, id: int) -> Optional[T]:
        """Find a record by ID"""
        return self.db.query(self.model).filter(self.model.id == id).first()
    
    async def find_all(self, limit: Optional[int] = None) -> List[T]:
        """Find all records with optional limit"""
        query = self.db.query(self.model)
        if limit:
            query = query.limit(limit)
        return query.all()
    
    async def create(self, **kwargs) -> T:
        """Create a new record"""
        instance = self.model(**kwargs)
        self.db.add(instance)
        self.db.commit()
        self.db.refresh(instance)
        return instance
    
    async def update(self, id: int, **kwargs) -> Optional[T]:
        """Update a record by ID"""
        instance = await self.find_by_id(id)
        if not instance:
            return None
        
        for key, value in kwargs.items():
            if hasattr(instance, key):
                setattr(instance, key, value)
        
        self.db.commit()
        self.db.refresh(instance)
        return instance
    
    async def delete(self, id: int) -> bool:
        """Delete a record by ID"""
        instance = await self.find_by_id(id)
        if not instance:
            return False
        
        self.db.delete(instance)
        self.db.commit()
        return True
    
    async def count(self) -> int:
        """Count total records"""
        return self.db.query(self.model).count()
